packages/realm/src/Dictionary.ts

```diff
@@ -322,3 +317,165 @@ export class Dictionary<T = unknown> extends Collection<string, T, [string, T],
     );
   }
 }
+
+/**
+ * Accessor for getting and setting items in the binding collection, as
+ * well as converting the values to and from their binding representations.
+ * @internal
+ */
+export type DictionaryAccessor<T = unknown> = TypeHelpers<T> & {
+  get: (dictionary: binding.Dictionary, key: string) => T;
+  set: (dictionary: binding.Dictionary, key: string, value: T) => void;
+  snapshotGet: (snapshot: binding.Results, index: number) => T;
+};
+
+type DictionaryAccessorFactoryOptions<T> = {
+  realm: Realm;
+  typeHelpers: TypeHelpers<T>;
+  isEmbedded?: boolean;
+  isMixedItem?: boolean;
+};
+
+/** @internal */
+export function createDictionaryAccessor<T>(options: DictionaryAccessorFactoryOptions<T>): DictionaryAccessor<T> {
+  return options.isMixedItem
+    ? createDictionaryAccessorForMixed<T>(options)
+    : createDictionaryAccessorForKnownType<T>(options);
+}
+
+function createDictionaryAccessorForMixed<T>({
+  realm,
+  typeHelpers,
+}: Pick<DictionaryAccessorFactoryOptions<T>, "realm" | "typeHelpers">): DictionaryAccessor<T> {
+  return {
+    get: (...args) => getMixed(realm, typeHelpers, ...args),
+    set: (...args) => setMixed(realm, typeHelpers.toBinding, ...args),
+    snapshotGet: (...args) => snapshotGetMixed(realm, typeHelpers, ...args),
+    ...typeHelpers,
+  };
+}
+
+function createDictionaryAccessorForKnownType<T>({
+  realm,
+  typeHelpers: { fromBinding, toBinding },
+  isEmbedded,
+}: Omit<DictionaryAccessorFactoryOptions<T>, "isMixed">): DictionaryAccessor<T> {
+  return {
+    get: (...args) => getKnownType(fromBinding, ...args),
+    set: (...args) => setKnownType(realm, toBinding, !!isEmbedded, ...args),
+    snapshotGet: (...args) => snapshotGetKnownType(fromBinding, ...args),
+    fromBinding,
+    toBinding,
+  };
+}
+
+function getKnownType<T>(fromBinding: TypeHelpers<T>["fromBinding"], dictionary: binding.Dictionary, key: string): T {
+  return fromBinding(dictionary.tryGetAny(key));
+}
+
+function snapshotGetKnownType<T>(
+  fromBinding: TypeHelpers<T>["fromBinding"],
+  snapshot: binding.Results,
+  index: number,
+): T {
+  return fromBinding(snapshot.getAny(index));
+}
+
+function getMixed<T>(realm: Realm, typeHelpers: TypeHelpers<T>, dictionary: binding.Dictionary, key: string): T {
+  const value = dictionary.tryGetAny(key);
+  if (value === binding.ListSentinel) {
+    const accessor = createListAccessor<T>({ realm, typeHelpers, isMixedItem: true });
+    return new List<T>(realm, dictionary.getList(key), accessor) as T;
+  }
+  if (value === binding.DictionarySentinel) {
+    const accessor = createDictionaryAccessor<T>({ realm, typeHelpers, isMixedItem: true });
+    return new Dictionary<T>(realm, dictionary.getDictionary(key), accessor) as T;
+  }
+  return typeHelpers.fromBinding(value) as T;
```

Comment:
For the way we're using symbols it would work. Given that the conditions are all just based on the value it may be preferable to use switch you like said. In that case, I'll have a look at the other places where I do the same thing 👍